docs/proposals/scaled-prometheus-pipeline/scaled-prometheus-pipeline.md

```diff
@@ -0,0 +1,139 @@
+Current metrics pipeline diagram:
+
+![currentMetricsPipeline.png](./currentMetricsPipeline.png)
+Since the pushgateway simply works as a temporary holder of metrics for prometheus to scrape, it does not currently work with a multiple-server model. If two prometheus servers scrape a pushgateway, there is no way to send only a subset of metrics to each server. This could be done of course, but it would add complexity to the pushgateway which I want to keep as simple as possible. We will need to deploy multiple pushgateways (one per prometheus server) and push a subset of metrics to each pushgateway. 
+
+![newMetricsPipeline.png](./newMetricsPipeline.png)
+
+To avoid making performance worse, we of course need to ensure that for each series, it is always sent to the same pushgateway and prometheus server. To do that, we need to implement some form of hashing. Preferably, this would all be done on the controllers, to avoid having to introduce a new “distributor” component. 
+
+### Distributing Metrics
+
+On prod, out of the ~105k metric series, only about 1.5k do not belong to a gateway. The average number of metric series per gateway is only ~350. This makes `gatewayID` a useful key to hash on. We simply map each `gatewayID` to a specific pushgateway, and all non-gateway metrics also to a single specific pushgateway. Even a very simple implementation with this method will ensure that no server has more than about 1000 series more than average. Importantly, all of this work can be done on the controller in the `Exporter`. 
+
+**Discussion:**
+While hashing on the controller provides an easy solution which is already scalable, it may be more architecturally sound to instead shift this responsibility to a load balancer in front of the pushgateways. Unfortunately, we cannot use a naive load-balancer since the metrics must be directed to the correct gateway. It would be possible to modify the pushgateway as a load balancer with this functionality. Simply configure it to hash the metrics as they come in and send them to the correct pushgateway to be scraped. However we then are left with another bottleneck to deal with. For this reason, I believe hashing on the controllers is the best option.
+
+### Dealing with Rescaling
+
+When a user decides to rescale their deployment (either up or down), series will inevitably be mapped to different pushgateways. This will mean that series will be scraped by different prometheus servers as well. Note that this will not cause any issues when querying/alerting as Thanos can handle the same series existing across multiple servers. The only effect this will have is some temporary non-optimal distribution of series since some will exist in two locations. Once the local server retention time has passed, everything will be back to normal, but even during that period we expect no noticeable impact to performance or user experience.
```

Comment:
Scaling up is probably less of an issue because you could send brand new series to the new prometheus servers.  It's more on the scaling down where you would have to redistribute series to existing servers.  My guess is that you won't scale down though, so probably not an issue.

Reply:
Consider the situation where you scale up because you're noticing performance degradation in the pushgateways (too much load). In this case you wouldn't just want to send new series to new gateways, but rather you'd want to redistribute all series evenly amongst the pushgateways. Additionally, since the hash should only be dependent on the `gatewayID` and the number of pushgateways, we wouldn't be able to send "new" series to the newly scaled up pushgateways.